plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,504 @@
+import           Language.PlutusIR.Analysis.Dependencies
+
+import           Control.Lens
+import           Control.Monad.Reader
+import           Control.Monad.RWS
+import           Control.Monad.State
+
+import qualified Language.PlutusCore                     as PLC
+import qualified Language.PlutusCore.Name                as PLC
+
+import qualified Algebra.Graph.AdjacencyMap              as AM
+import qualified Algebra.Graph.AdjacencyMap.Algorithm    as AM
+import qualified Algebra.Graph.NonEmpty.AdjacencyMap     as AMN
+
+import           Data.Either                             (fromRight)
+import           Data.Function                           (on)
+import qualified Data.IntMap                             as IM
+import qualified Data.Map                                as M
+import qualified Data.Set                                as S
+
+import qualified Data.List.NonEmpty                      as NE
+
+{- Note [Float algorithm]
+
+The goal of this PIR->PIR transformation is to float lets to their closest-surrounding lambda/Lambda abstraction and
+potentially merge adjacent lets into let-groups for
+better readability of the program's IR (and potentially to reveal any later code optimizations).
+
+Each let-binding is moved right underneath its closest-surrounding lambda/Lambda abstraction (or at toplevel in case of no surrounding lambdas).
+Then all let-bindings belonging to the same lambda/Lambda are merged into 1 or more let-groups. The order and the cardinality of
+the let-groups under the lambda/Lambda is determined by the dependency-graph of the program's IR.
+
+Specifically the algorithm is done in two passes:
+
+1st-pass: Traverses an IR term/program and for each let, mark the location of its closest-surrounding lambda (or Top if no lambda),
+and store it into a mapping of let-name=>lambda-location. We call the closest-surrounding lambda-location the "Rank" of a let.
+
+2nd-pass: Cleans the term from all its lets (see `removeLets` function) and traverses
+the clean-term to place them back again, this time  directly under their "Rank".
+During traversal, when a lambda-location of interest is reached (i.e. Rank),
+the algorithm consults the dependency graph to figure (a) the groupings of lets
+into let-groups based on their (mutual) recursivity (by utilizing the SCCs of the depgraph)
+and (b) the correct code-generation *order* of the let-groups (by utilizing the topsort of the SCCs).
+As an optimization, the algorithm merges adjacent *nonrec* let-groups into a single-group for better readability; this is possible
+because letnonrec in PlutusIR is linearly scoped.
+
+Invariants for algorithm:
+
+a) Does not float lets outside surrounding lambda/Lambda abstractions. This has the effect of preserving the lazy/strict semantics of the original program.
+b) Does not break scoping; only uses the `linear scoping` feature of let-nonrec  to merge adjacent letnonrecs
+c) It will not demote a letrec to a letnonrec.
```

Comment:
Y @michaelpj  is right, the prepass will not help in these corner cases, because the dependency graph is the source of truth. But in general and independently this demotion of letrecs to letnorecs might be an interesting optimization.

Reply:
After running an experiment to use the depgraph to calculate the demotions, it seems 1) that the depgraph is not adequate enough to achieve this and 2) it is perhaps better to do this demoting transformation *after* the DCE